src/gradient_aware_harmonisation/convergence.py

```diff
@@ -0,0 +1,335 @@
+    #     np.finfo(np.float64).max,
+    # ]
+    # """Domain of spline"""
+
+    @overload
+    def __call__(self, x: int | float) -> int | float: ...
+
+    @overload
+    def __call__(self, x: NP_FLOAT_OR_INT) -> NP_FLOAT_OR_INT: ...
+
+    @overload
+    def __call__(self, x: NP_ARRAY_OF_FLOAT_OR_INT) -> NP_ARRAY_OF_FLOAT_OR_INT: ...
+
+    def __call__(
+        self, x: int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT
+    ) -> int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT:
+        """
+        Evaluate the spline at a given x-value
+
+        Parameters
+        ----------
+        x
+            x-value
+
+        Returns
+        -------
+        :
+            Value of the spline at `x`
+        """
+
+        # compute weight (here: gamma) according to a cosine-decay
+        def calc_gamma_rising_derivative(
+            x: int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT,
+        ) -> int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT:
+            """Get cosine-decay derivative"""
+            # compute derivative of gamma according to a cosine-decay
+            angle = (
+                np.pi * (x - self.initial_time) / (self.final_time - self.initial_time)
+            )
+            gamma_decaying_derivative = -0.5 * np.sin(angle)
+
+            return gamma_decaying_derivative
+
+        if not isinstance(x, np.ndarray):
+            if x <= self.initial_time or x >= self.final_time:
+                return 0.0
+
+            gamma_rising_derivative = calc_gamma_rising_derivative(x)
+
+            # The weighted sum for computing the harmonised AND converged
+            # function has the form: "gamma * harmonised + (1-gamma) * convergence".
+            # Depending on which product we want to compute (LHS or RHS of sum),
+            # we need gamma or 1-gamma, therefore we include the following condition
+            # in all our return statements.
+            if self.apply_to_convergence:
+                return -gamma_rising_derivative
+
+            return gamma_rising_derivative
+
+        # apply decay function only to values that lie between harmonisation
+        # time and convergence-time
+        x_decay = np.where(np.logical_and(x > self.initial_time, x < self.final_time))
+        gamma_rising_derivative = np.zeros_like(x, dtype=np.floating)
+        gamma_rising_derivative[x_decay] = calc_gamma_rising_derivative(x[x_decay])
+
+        if self.apply_to_convergence:
+            return -gamma_rising_derivative
+        return gamma_rising_derivative
+
+    def derivative(self) -> CosineDecaySplineHelperDerivative:
+        """
+        Calculate the derivative of self
+
+        Returns
+        -------
+        :
+            Derivative of self
+        """
+        return CosineDecaySplineHelperDerivative(
+            initial_time=self.initial_time,
+            final_time=self.final_time,
+            apply_to_convergence=self.apply_to_convergence,
+        )
```

Comment:
```suggestion
        raise NotImplementedError
```

We'd have to be a bit careful to make this work, no?